R/po_compile.R

```diff
@@ -0,0 +1,62 @@
+#' Compile `.po` files to `.mo`
+#'
+#' This function compiles the plain text `.po` files that translators work with
+#' into the binary `.mo` files that are installed with packages and used for
+#' live translations.
+#'
+#' @param dir Path to package root directory.
+#' @param package Name of package. If not supplied, read from `DESCRIPTION`.
+#' @param lazy If `TRUE`, only `.mo` functions that are older than `.po`
+#'   files be updated
+#' @param verbose If `TRUE`, print information as it goes.
+po_compile = function(dir = ".", package = NULL, lazy = TRUE, verbose = TRUE) {
+  po_metadata <- get_po_metadata(dir = dir, package = package)
+  dir_create(dirname(po_metadata$mo))
+
+  if (lazy) {
+    po_metadata <- po_metadata[is_outdated(po_metadata$po, po_metadata$mo)]
+  } else {
+    # Clear out all older translations
+    mo_dirs <- dir(file.path(dir, "inst", "po"), full.names = TRUE)
+    to_delete <- mo_dirs[!basename(mo_dirs) %in% c(po_metadata$language, "en@quot")]
+
+    for (dir in to_delete) {
+      if (verbose) messagef(
+        "Found a compiled translation for %s at %s, but no corresponding .po; deleting",
+        basename(dir), dirname(dir)
+      )
+      unlink(dir, recursive = TRUE)
+    }
+  }
+
+  for (ii in seq_len(nrow(po_metadata))) {
+    row <- po_metadata[ii]
+    if (verbose) messagef("Recompiling '%s' %s translation", row$language, row$type)
+    run_msgfmt(row$po, row$mo, verbose = verbose)
+  }
+
+  return(invisible())
+}
+
+
+get_po_metadata <- function(dir = ".", package = NULL) {
+  if (is.null(package)) {
+    package <- get_desc_data(dir)[["Package"]]
```

Comment:
should we add `fields` to the `get_desc_data` signature?

Reply:
We could, but I don't think it simplifies the code much, since you'd still need `get_desc_data(dir, "Package")[[1]]` or similar. Maybe revisit after reading the style from the DESCRIPTION?

Reply:
We could, but I don't think it simplifies the code much, since you'd still need `get_desc_data(dir, "Package")[[1]]` or similar. Maybe revisit after reading the style from the DESCRIPTION?

Reply:
Hmm I'm not sure -- the `get_desc_data` output is already a vector, so it's just a matter of whether we keep the names (to do `[[`), right?

Extended `get_desc_data` in #255

Reply:
Oh sorry, I forgot it's a character vector and not a list.